### PR TITLE
Add support for "true"/"false" intra doc link

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/rust-lang/rust/issues/74515.

Explanations of what I did here: instead of checking more globally in primitive types, I think that only `true` and `false` values should be transformed into a link to `bool`. Others are either numbers or strings and I don't think we want to link all of them "generally". So in here, in case `true` and `false` didn't match anything (maybe a module was here, we don't know?), we then link them to the `bool` type.

r? @jyn514 

cc @rust-lang/rustdoc

### CODE REVIEW

**Comment by jyn514** [comment]
How do we know these are the only primitive values that have keywords? I'm ok with this workaround but I want to be sure we didn't miss anything.

**Comment by jyn514** [comment]
Ok, there is a list of keywords at https://doc.rust-lang.org/std/index.html#keywords.

Do we want to support all of those, like `as` and `dyn`? Or just the keywords that are types?

**Comment by jyn514** [comment]
I think we should at least support `enum`, `union`, and `trait`.

**Comment by GuillaumeGomez** [comment]
`keyword` != `primitive`. In here we use `values` which are in fact of primitive type. `enum`, `union` and `trait` aren't. Or maybe I missed your point?

**Comment by GuillaumeGomez** [comment]
CI failure:

```
warning: spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/autocfg/1.0.0/download`, got 504
error: failed to download from `https://crates.io/api/v1/crates/libc/0.2.74/download`
```

**Comment by Mark-Simulacrum** [comment]
It seems a bit odd to link to the bool type for these rather than https://doc.rust-lang.org/nightly/std/keyword.true.html and https://doc.rust-lang.org/nightly/std/keyword.false.html -- if we're going to autolink keywords, we should link to the keyword documentation, I think.

**Comment by GuillaumeGomez** [comment]
Oh I see, we had two different things in mind. Then we could try to do this based on the doc aliases directly, don't you think? Or should we keep a list based on the official keywords and only link using those? The second solution would allow a much simpler fix.

**Comment by jyn514** [comment]
I think linking by doc-aliases would be the best approach - that would link to all keywords, right?

The issue with special-casing keywords is that now we have to hard-code either stable or nightly; if we use resolve for it then it will always be the appropriate version.

**Comment by GuillaumeGomez** [comment]
Ok, I'll update the PR with this change.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -747,30 +747,40 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
...
-                            candidates.present_items().next().unwrap()
-                        } else if len == 2 && is_derive_trait_collision(&candidates) {
-                            candidates.type_ns.unwrap()
+                            if path_str == "true" || path_str == "false" {
+                                match PRIMITIVES.iter().find(|(ty, _)| *ty == "bool") {
+                                    Some((_, res)) => (*res, Some("bool".to_owned())),
+                                    None => {
+                                        resolution_failure(cx, &item, path_str, &dox, link_range);
+                                        continue;
+                                    }
```
Comment (pickfire) [inline src/librustdoc/passes/collect_intra_doc_links.rs:756]: Should this be always checking in a loop? Shouldn't we get `bool` only once or make this static?

File: src/test/rustdoc/intra-doc-link-true-false.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#![deny(broken_intra_doc_links)]
+#![crate_name = "foo"]
+
+// ignore-tidy-linelength
+
+// @has foo/index.html
+// @has - '//*[@id="main"]//a[@href="https://doc.rust-lang.org/nightly/std/primitive.bool.html"]' 'true'
+// @has - '//*[@id="main"]//a[@href="https://doc.rust-lang.org/nightly/std/primitive.bool.html"]' 'false'
+
+//! A `bool` is either [`true`] or [`false`].
```
Comment (pickfire) [inline src/test/rustdoc/intra-doc-link-true-false.rs:10]: Shouldn't we also test ```[`bool`]```?

Comment (GuillaumeGomez) [reply src/test/rustdoc/intra-doc-link-true-false.rs:10]: It's already tested somewhere else, like all primitive types that we support.

**Comment by Manishearth** [comment]
I think we should link to the keywords  here.

But I'm kinda skeptical of doing this at all, these are keywords and not paths. Seems ok for true and false but maybe not so much for other keywords. I also see a very low utility for it, and because they're not paths, I don't expect people to reach for them aside from the bool literals (which behave kinda like paths, at least)

**Comment by jyn514** [comment]
FWIW I'm ok with just not implementing this, this is definitely out of the scope of the original RFC. Up to @GuillaumeGomez if he wants to go through with the PR since he already wrote a lot of it.

If we do implement it though, I think it would be nice to make it work for all keywords.

**Comment by GuillaumeGomez** [comment]
Honestly I'm not sure it's worth it so let's just keep it simple. I'll update to link to the keyword pages though.

**Comment by Manishearth** [comment]
I'm very against linking to all keywords, the only reason I'm somewhat okay with this one is because `true` and `false` behave _kind of_ like paths; they operate in path positions (unlike most other keywords).

**Comment by GuillaumeGomez** [comment]
Yes we agree, only keywords for `true` and `false`. ;)

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by Manishearth** [comment]
No that was a response to @jyn514 who wanted us to do all keywords if we do true/false.

**Comment by GuillaumeGomez** [comment]
Oh I see, well, we can talk about it at a later time if enough people are interested. But currently, not very convinced.

**Comment by ollie27** [comment]
`bool` is basically an `enum` so having `true` and `false` link to the page for `bool` the same way `None` and `Some` link to the page for `Option` makes some sense. I don't think linking to the keyword pages is better but then again I'm not a fan of the keyword pages in general. How often are we expecting people to use this, and for what?

**Comment by GuillaumeGomez** [comment]
I expect people to use them as soon as the intra-doc links hit stable. For the exact numbers, we'll have to wait for some time.

**Comment by jyn514** [comment]
I originally filed the issue because @poliorcetics tried to use these in https://github.com/rust-lang/rust/pull/73618 and found that they didn't work. But documenting keywords is something only std is likely to do, which is why I'm ok with not implementing it.

**Comment by ollie27** [comment]
> tried to use these in #73618 and found that they didn't work.

If that's really the only use case then this definitely isn't worth the complexity.

**Comment by Manishearth** [comment]
@bors r-

two team members are against it at this point. Maybe we should hold off unless there's strong motivation?

**Comment by GuillaumeGomez** [comment]
I don't have strong feelings here so we can close it if everyone agrees?

**Comment by jyn514** [comment]
Sounds good to me.

**Comment by jyn514** [comment]
This came up again in https://github.com/rust-lang/rust/pull/75567, which has 38 different links to `true` and `false` ... what do you think about only linking `true` and `false` and no other keywords? I took a bit of a strong position last time 😆

cc @poliorcetics

**Comment by GuillaumeGomez** [comment]
Well, if enough people are interested, please reopen then. :laughing:

**Comment by jyn514** [comment]
I can't re-open because you deleted the branch :/
deleted

**Comment by GuillaumeGomez** [comment]
Restored the branch has been. :3

**Comment by ollie27** [comment]
> what do you think about only linking `true` and `false` and no other keywords?

What would they link to? I think linking to the [`bool`](https://doc.rust-lang.org/nightly/std/primitive.bool.html) page makes sense and would be more useful than the keyword pages which don't contain much.

**Comment by Manishearth** [comment]
I agree with Ollie here

**Comment by jyn514** [comment]
Linking to `bool` sounds fine to me, that's actually where they go currently: https://github.com/rust-lang/rust/pull/75567/files#diff-3935db3bd98ce45bbc780286b8d11c78R212

**Comment by poliorcetics** [comment]
`true` and `false` would be very nice to have for documenting rust itself (I expected them to behave like `Ok` or `Some`) but if it needs a lot of special-casing code, don't bother. They shouldn't be used too much outside `std`, `core` or `alloc` so we should ensure the benefits outweighs the cost (though that claim is completely unverified).

**Comment by jyn514** [comment]
> if it needs a lot of special-casing code, don't bother

If we're just linking to bool, the code can be simplified quite a bit, @guillaumegomez just needs to add an alias next to `bool`: https://github.com/rust-lang/rust/blob/master/src/librustdoc/passes/collect_intra_doc_links.rs#L1222

> I expected them to behave like Ok or Some

Well, they can't really because `Ok` and `Some` are documented on the same page as `Result`, but all the keywords are on separate pages. Does linking to `bool` sound like the right behavior?

**Comment by poliorcetics** [comment]
Oh that's nice if it doesn't ask for a lot of code.

Yeah, linking to bool is better I think.

**Comment by jyn514** [comment]
Opened https://github.com/rust-lang/rust/pull/75652 which links `true` and `false` to `bool`. It's a little simpler because it doesn't have to deal with keywords.

@GuillaumeGomez do you mind if I close this PR (again :P)?

**Comment by GuillaumeGomez** [comment]
Too late, I already did! :p

**Pull request closed without merging**